src/controllers/vdeployment_controller/proof/predicate.rs

```diff
@@ -1,47 +1,503 @@
+        &&& request.namespace == vd.metadata.namespace.unwrap()
+        &&& request.owner_ref == vd.controller_owner_ref()
+        &&& VReplicaSetView::unmarshal(request.obj).is_ok()
+        &&& s.resources().contains_key(key)
+        // step-specific update content
+        &&& req_vrs.metadata.owner_references_contains(vd.controller_owner_ref())
+        &&& req_vrs.spec.replicas.unwrap_or(1) == n
+    }
+}
+
+pub open spec fn req_msg_is_pending_get_then_update_req_in_flight_with_replicas(
+    vd: VDeploymentView, controller_id: int, req_msg: Message, n: int
+) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        &&& Cluster::pending_req_msg_is(controller_id, s, vd.object_ref(), req_msg)
+        &&& s.in_flight().contains(req_msg)
+        &&& req_msg_is_get_then_update_req_with_replicas(vd, controller_id, req_msg, n)(s)
+    }
+}
+
+pub open spec fn pending_get_then_update_req_in_flight_with_replicas(
+    vd: VDeploymentView, controller_id: int, n: int
+) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        let req_msg = s.ongoing_reconciles(controller_id)[vd.object_ref()].pending_req_msg.get_Some_0();
+        &&& Cluster::pending_req_msg_is(controller_id, s, vd.object_ref(), req_msg)
+        &&& s.in_flight().contains(req_msg)
+        &&& req_msg_is_get_then_update_req_with_replicas(vd, controller_id, req_msg, n)(s)
+    }
+}
+
+// a weaker version of coherence between local cache and etcd
+// - only need the key to be in etcd and corresponding objects can pass the filter
+// - so current_state_matches can be reached by sending get-then-update request
+// this predicate holds since AfterListVRS state
+// TODO: simplify this may help to reduce flakiness
+pub open spec fn local_state_is_consistent_with_etcd(vd: VDeploymentView, controller_id: int) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        let vds = VDeploymentReconcileState::unmarshal(s.ongoing_reconciles(controller_id)[vd.object_ref()].local_state).unwrap();
+        &&& forall |i| 0 <= i < vds.old_vrs_list.len() ==> {
+            // the get-then-update request can succeed
+            &&& #[trigger] vds.old_vrs_list[i].well_formed()
+            &&& #[trigger] vds.old_vrs_list[i].metadata.namespace == vd.metadata.namespace
+            &&& #[trigger] vds.old_vrs_list[i].metadata.owner_references_contains(vd.controller_owner_ref())
+            // obj in etcd exists and is owned by vd
+            &&& s.resources().contains_key(vds.old_vrs_list[i].object_ref())
+            &&& ({
+                let etcd_obj = s.resources()[vds.old_vrs_list[i].object_ref()];
+                let etcd_vrs = VReplicaSetView::unmarshal(etcd_obj).unwrap();
+                // can pass list_vrs_obj_filter
+                &&& etcd_obj.kind == VReplicaSetView::kind()
+                &&& etcd_obj.metadata.namespace == vd.metadata.namespace
+                // can pass objects_to_vrs_list
+                &&& VReplicaSetView::unmarshal(etcd_obj).is_ok()
```

Comment:
You don't have to carry `VReplicaSetView::unmarshal(etcd_obj).is_ok()` in this predicate. There should be an invariant stating that all VRS objects in etcd can be unmarshalled correctly.